### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by carlm0101**
background.js:

Create "call" exactly once at Install time. This eliminates the harmless error message (when "Collect errors" is ON) about a duplicate context menu item.

background.js:

In the case where phoneNumberRegex.test(selectedText) fails to find a match, run  encodeURIComponent() on selectedText. Solves problem with leading '+' (a special character) in selectedText being passed in URL without encoding and thus not appearing in REQUEST parameter in destination script.

### CODE REVIEW

**Comment by carlm0101** [comment]
I think my addition of encodeURIComponent() in background.js creates a problem with international phone numbers with a '+' prefix and the tel: protocol.

With html: protocol the '+' at the start of the phone number must be encoded, which encodeURIComponent() does.  I normally generate html: URIs w/ TelLinker so adding encodeURIComponent() fixed my problem.

I just tried generating tel: URIs w/ TelLinker and discovered that the encoded '+' is not decoded by the tel: protocol handler I used. In RFC 3966 (https://tools.ietf.org/html/rfc3966) I found the following:

"If the reserved characters  "+", ";", "=", and "?" are used as delimiters between components of
   the "tel" URI, they MUST NOT be percent encoded."

Thus the tel: protocol handler I used is not at fault.

At present I'm not sure what to do to handle the conflicting requirements of the http: and tel: protocols.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI  

"Note that encodeURI by itself cannot form proper HTTP GET and POST requests, such as for XMLHTTPRequests, because "&", "+", and "=" are not encoded, which are treated as special characters in GET and POST requests. encodeURIComponent, however, does encode these characters."

Thus simply changing encodeURIComponent() to encodeURI() is not a solution.

One approach, which I have seen in ADAT (http://adat.tttelecom.nl/) and Chrome extension bitree.ru Asterisk Click2Call, is to incorporate user-defined phone number rewriting rules (regex) in the program. In this case a rewriting rule would replace /^\+/ with '011' or '00', whichever applies in the user's country. This eliminates the need to pass '+' to the protocol handler.

Let me know what you think.

**Comment by mjones2219** [comment]
I suppose I should have vetted the code first but I don’t currently use the extension and it seemed harmless. My bad. The easiest, quickest solution is to just throw something in there to check for tel and not do the encode. This means moving where the encoding occurs. It was always the plan to allow user created regex that way it worked for more than just US numbers, but the extension stopped being useful to me before I got around to it. I’ll get around to attempting to fix it tomorrow.

**Comment by carlm0101** [comment]
I just put a new background.js on my contrib branch which fixes (tested) the bug in my previous background.js. There's a detailed comment.
I just noticed that you have put up v0.5 for testing. Haven't tried it yet, but noticed that there's no longer any text selection.
Do you anticipate any further releases of v0.4? Should I create a pull request?

**Comment by mjones2219** [comment]
Test version 5 to see if it works for you with http/https. I’ve already published it. I anticipate merging this branch into the master branch once I confirm I haven’t introduced any new bugs. This version also allows custom regex. I did some testing and the + symbol is an allowed symbol for http/https. The encodeURI won’t touch it. If you’re passing a plus symbol that needs encoring, and it’s hard coded into your tel link format, have you considered encoding it in the settings instead? Although I’m not sure if it was the + you were originally worrying about.

**Comment by mjones2219** [comment]
Hmm. I also just realized I removed the context action and forgot to add it back. I'll add that back in later.

If you want to build out the UI more to differentiate web requests vs. protocol requests that would honestly probably work better, but I was going for a bare-bones just-works kind of approach, even though it means it might require someone a more technical to get it to do things other than its default behavior.

**Comment by mjones2219** [comment]
Actually, the most recent commit includes something that should work fine without you having to encode it in the settings. I'm testing it as an unpacked extension right now.

**Comment by carlm0101** [comment]
That '+' is a real problem. I can't hard code it in the link format. It's used as a flag that the phone number requires an international direct dial prefix, and not all calls are international. Both my http: and tel: scripts replace '+' with '011'.  At the PBX, presence/absence of '011' determines whether the call goes to the international/domestic route. RFC 3966 states that the '+' is always required in tel:, but most everyone ignores that for good reason. Note: In some countries it's '00' rather than '011'.

Now that I've fixed my over-zealous encoding in background.js I'm very happy with the way v0.4 is working. I don't mind that it doesn't generate links for non-NANP numbers as I rarely dial such numbers and having to first select the text on rare occasions doesn't bother me.

I don't feel confident in being able to create a regex that correctly recognizes all international numbers while rejecting everything that is not a phone number. If you think you have such a regex let me know and I'll find a web page with an international number that it can't recognize. :-)

Worth mentioning: Adding user-defined regex may result in a lot more support requests.

RE adding protocol selection to the UI: I think the logic that I added to background.js makes that unnecessary.

**Comment by mjones2219** [comment]
I'll likely ignore support requests .. I don't use the extension anymore.. I know that isn't a good reason but it is what it is haha.

I probably overthought this, but this is what I ended up adding. I have no idea if this ultimately produces the same thing you were doing or not:

`if (settings.telLinkFormat.startsWith("http")) {
            var numberString = settings.telLinkFormat.substring(0, settings.telLinkFormat.indexOf('{')) + selectedText;
            var url = new URL(numberString);
            //this feels stupidly hacky but it seems to get the job done so.. lets do it?
            var urlString = url.protocol + "//" + url.hostname + url.pathname;
            numberString = numberString.replace(urlString, "").replace("?", "");
            var parts = numberString.split("=");
            if (parts.length > 0) {
                urlString += "?";
                for (var i = 0; i < parts.length; i+=2) {
                    urlString += parts[i] + "=" + encodeURIComponent(parts[i+1]);
                }
            }
            chrome.tabs.update(tab.id, { url: urlString });
        } else {
            chrome.tabs.update(tab.id, { url: settings.telLinkFormat.substring(0, settings.telLinkFormat.indexOf('{')) + selectedText });
}`

**Comment by mjones2219** [comment]
That is ugly as hell...

Here is the commit: https://github.com/mjones2219/TelLinker/commit/e3edb7e32a2cfc23a8a0a991e176e3c0a778d28c

It tries to parse out any variables and encode their values, which is how a GET url should be setup. How were you getting the +1 replaced with a 011? I don't think the regex I had in there would work on international numbers - I only ever intended it for U.S. numbers. Are you using your own forked version with different regex?

**Comment by carlm0101** [comment]
Replacement of '+' with '011' happens in my scripts for tel: and http: , not in my background.js. That's why it's essential that the '+' gets passed to the script (%2B for http: , '+' for tel: ). No, your regex doesn't match international numbers, but that's OK with selected text because your logic on match failure passes the entire selection to the protocol handler, encoded or raw depending upon the protocol (with my mod).

In other words, with an international number in v0.4 the autopilot doesn't work and the user has to select, right-click, click context menu.  Doesn't bother me, but as I said I rarely place international calls.

As soon as v0.5 settles down a bit I'll give it a try.

**Comment by carlm0101** [comment]
Haven't tried v0.5 yet, but I did read through background.js and I see you are matching against multiple user-defined regex's. Hadn't realized that before. Obviates the need for the "mother of all regex's" that I had envisioned. Might require a lot of trial and error to get the right set of regex's for international numbers, though.

FYI, other approaches to international numbers that I have seen:

bitree.ru Asterisk Click2Call : one rather short regex to identify phone numbers. Haven't paid any attention to it and can't say how well it works, as identifying phone numbers is a premium (paid) feature. But you can see the regex in the UI without paying.

Telify always pops up an "Edit number" dialog, i.e. it leaves the heavy lifting to the user.

**Comment by mjones2219** [comment]
Might throw something like this into fiddle and see if it works. I never call international numbers but it might work. I think to properly handle iterating through regex I will have to double back and make the format links pair with the regex otherwise you’re really limited to a single format style meaning if you’re trying to do US numbers + international you’re going to struggle to make it work with a single format entry, unless you just use the {0} index. The version that is posted will encode + as long as it’s a proper GET url, ie, follows the ?key=value format.  

^\+((?:9[679]|8[035789]|6[789]|5[90]|42|3[578]|2[1-689])|9[0-58]|8[1246]|6[0-6]|5[1-8]|4[013-9]|3[0-469]|2[70]|7|1)(?:\W*\d){0,13}\d$
